ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,524 @@
+
+The `CommitListener` we use for these components, expect entries especially positions to be committed, but internally in RAFT we commit indexes. This is the reason why we need some glue code to translate indexes to position's.
+
+In order to explain that more in detail, we first need a basic understanding how a RaftEntry looks like and what an index is, for that we can take a look at the following picture.
+
+![index](images/IndexedRaftEntry.png)
+
+RAFT entries are written to the log. Each entry contains an index, which can be seen as the identifier of such an entry. Indexes are incremented by one. There are multiple types of raft entries, here we will just concentrate on an raft entry which contain application entries. The `IndexedRaftEntry` can contain multiple application entries. The application entries are sorted in a raft entry. The raft entry knows the lowest and highest position, which references the first and last application entry. 
+
+In RAFT these entries are replicated via `AppendRequests` by the Leader. An `AppendRequest` looks like this:
+
+```
+AppendRequest {
+  long term
+  String leader
+  long prevLogIndex
+  long prevLogTerm
+  List<PersistedRaftRecord> entries
+  long commitIndex
+}
+```
+
+An [AppendRequest](https://github.com/camunda-cloud/zeebe/blob/develop/atomix/cluster/src/main/java/io/atomix/raft/protocol/AppendRequest.java) can contain multiple raft entries. If the quorum is reached for a certain index, the index can be committed. The last committed index is shared together with the raft entries via an `AppendRequest`.
+
+![leaderCommit](images/leaderCommit.png)
+
+If the Leader commits an index it can immediately call the RAFT commit listeners with the corresponding `IndexedRaftEntry`. This is necessary, since we need to translate the index of the committed entry into a committed position later. We can do this in the `AtomixLogStorage` via asking for the highest position, if the entry contains application entries. With this highest position we can set the commit position in our LogStream abstraction. The LogStream will then notify all our other commit listeners, which wait for the committed position.
+
+On the follower it looks similar, but it works a bit different.
+
+![followerCommit](images/followerCommit.png)
+
+As written above, the follower receives `AppendRequests`, which will contain the last committed index. The committed index, does need to correspond to the entries which are sent via the `AppendRequest`. In order to be able to translate the index to the position, we need the `IndexedRaftEntry`. For that the Follower has a separate reader, which is always forwarded to the next committed index to get the latest committed IndexedRaftEntry, which he then uses to call the RAFT `CommitListeners`. The rest stays the same as on the Leader side.
```

Comment:
As I said before, I'm not a fan of having another reader for the sake of commit listener. But I wouldn't see it as a blocker for now. We can always improve if we find a better solution.

Reply:
Agree. Alternative would be https://github.com/zeebe-io/enhancements/pull/16#discussion_r661751155

Reply:
I think we will go with the proposed solution in the comment, which allows to avoid a separate reader. I will rewrite this section.